include/boost/geometry/algorithms/detail/envelope/box.hpp

```diff
@@ -40,36 +42,124 @@ namespace detail { namespace envelope
+    std::size_t Dimension,
+    std::size_t DimensionCount,
+    typename CS_Tag
+>
+struct envelope_indexed_box
 {
-    template<typename BoxIn, typename BoxOut>
+    template <typename BoxIn, typename BoxOut>
     static inline void apply(BoxIn const& box_in, BoxOut& mbr)
     {
-        BoxIn box_in_normalized = detail::return_normalized<BoxIn>(box_in);
+        detail::indexed_point_view<BoxIn const, Index> box_in_corner(box_in);
+        detail::indexed_point_view<BoxOut, Index> mbr_corner(mbr);
+
+        detail::conversion::point_to_point
+            <
+                detail::indexed_point_view<BoxIn const, Index>,
+                detail::indexed_point_view<BoxOut, Index>,
+                Dimension,
+                DimensionCount
+            >::apply(box_in_corner, mbr_corner);
+    }
+};
 
-        geometry::transform(box_in_normalized, mbr);
+template
+<
+    std::size_t Index,
+    std::size_t DimensionCount
+>
+struct envelope_indexed_box<Index, 0, DimensionCount, spherical_equatorial_tag>
+{
+    template <typename BoxIn, typename BoxOut>
+    static inline void apply(BoxIn const& box_in, BoxOut& mbr)
+    {
+
+       // transform() does not work with boxes of dimension higher
+        // than 2; to account for such boxes we transform the min/max
+        // points of the boxes using the indexed_point_view
+        detail::indexed_point_view<BoxIn const, Index> box_in_corner(box_in);
+        detail::indexed_point_view<BoxOut, Index> mbr_corner(mbr);
+
+        // first transform the units
+        transform_units(box_in_corner, mbr_corner);
+
+        // now transform the remaining coordinates
+        detail::conversion::point_to_point
+            <
+                detail::indexed_point_view<BoxIn const, Index>,
+                detail::indexed_point_view<BoxOut, Index>,
+                2,
+                DimensionCount
+            >::apply(box_in_corner, mbr_corner);
     }
 };
 
+template
+<
+    std::size_t Index,
+    std::size_t Dimension,
+    std::size_t DimensionCount
+>
+struct envelope_indexed_box<Index, Dimension, DimensionCount, geographic_tag>
+    : envelope_indexed_box
+        <
+            Index, Dimension, DimensionCount, spherical_equatorial_tag
+        >
+{};
+
 
-template <typename CSTag>
+template
+<
+    typename std::size_t Dimension,
+    std::size_t DimensionCount,
+    typename CS_Tag
+>
 struct envelope_box
 {
     template<typename BoxIn, typename BoxOut>
     static inline void apply(BoxIn const& box_in, BoxOut& mbr)
     {
-        geometry::convert(box_in, mbr);
+        envelope_indexed_box
+            <
+                min_corner, Dimension, DimensionCount, CS_Tag
+            >::apply(box_in, mbr);
+
+        envelope_indexed_box
+            <
+                max_corner, Dimension, DimensionCount, CS_Tag
+            >::apply(box_in, mbr);
+    }
+};
+
+
+template <typename std::size_t DimensionCount>
+struct envelope_box<0, DimensionCount, spherical_equatorial_tag>
```

Comment:
In general, such templates taking the `Dimension` in Boost.Geometry works specifically on this Dimension. The way how the algorithm is designed and divided into templates is confusing for me. I also don't see a reason why the algorithm is separated into the `envelope_box` and `envelope_indexed_box` both dispatched with `Dimension`. It makes very simple thing quite complicated. And this simple thing is:
1. for cartesian:
    1.1 expand_coordinates<0, DimCount>(...)
2. for spherical and geographic
    2.1 normalize_coordinates_0_and_1()
    2.2 transform units_of_coordinates_0_and_1()
    2.3 expand_coordinates<2, DimCount>(...)
The dispatching with `Dimension` is needed only on the `expand_coordinates` step. The `envelope_box` must be only dispatched with CSTag. If in specializations done for coordinate systems those such utilities as mentioned above were used it'd be clear what's happening here. Those utilities could still be reusable, taking ranges of dimensions.

Or am I missing something and there is some usage I don't see?